Design/histogram_type.md

```diff
@@ -0,0 +1,251 @@
+   // Errors are dealt with implicitly unless specified otherwise
+   const auto &errors = histogram.countStandardDeviations();
+   // errors[0] should now be 4
+   ```
+
+4. As shown in the previous example, in many cases you do not have to deal with errors yourself.
+   As long as the standard deviation is simply the square root of the counts it will be computed and propagated under the hood.
+
+5. `Variances` and `StandardDeviations` can be automatically converted, you do not have to deal with it yourself.
+
+
+#### Details
+
+- The different low-level types have been introduced in the previous section.
+  To avoid highly duplicate code and tests, such as for the four `Variance` types, these are implemented using the curiously recurring template pattern (CRTP) which provides something like "mixins".
+- Jon pointed out that it may make most sense to default to storing data as `Frequecies`, as OpenGenie did by default.
+  Many operations could probably done in this representation, and furthermore the conversion to `Counts` is cheaper, since it includes a multiplication instead of addition.
+- The current copy-on-write (COW) mechanism seems useful and should be preserved.
+  - The `typedef` `MantidVecPtr` for a `cow_ptr` is useless and confusing and should be removed.
+  - We should consider extending the COW mechanism and make sure sharing is preserved more than it is now. For example, what can we do to maintain sharing of `X` for all histograms in a workspace where `X` is modified in an identical way? The COW mechanism as it is now should break down in that case. Probably this needs to be handled at the workspace/algorithm level, and cannot be dealt with this here? Having parametric `X` could circumvent that in certain cases (see outlook).
+
+#### Interface
+
+- The interface of `Histogram` can be (partially) forwarded by `ISpectrum` and `MatrixWorkspace`, similarly to how it is implemented currently for, e.g., `ISpectrum::dataX` and `MatrixWorkspace::dataX`.
+
+- The main challenge in the design of the `Histogram` interface is the way how we deal with references.
+  If the internal data storage mode of, say, the Y-data is not known, it is not trivial to provide `Counts &counts()` (as a partial equivalent to the current `std::vector<double> &dataY()`), since there may be nothing to reference to.
+  - One way of dealing with this is a conversion of the underlying type on write, e.g.,
+
+    ```cpp
+    // Default storage mode is `Frequencies`.
+    Histogram hist;
+    // Const accessors return by value:
+    // const Counts Histogram::constCounts() const;
+    const auto &counts = hist.constCounts();
+    // Non-const accessors return by reference and converts internal storage mode:
+    // Counts &Histogram::counts();
+    auto &writableCounts = hist.counts();
+    // modify writableCounts
+    ```
+
+  - Alternatively, we could eliminate the option of getting non-const references, and instead do everything via setters, e.g.,
```

Comment:
@martyngigg The internal data representation is not necessary equal to what the getter access, e.g., data could be stored as `Frequencies`. We then cannot return a reference to `Counts` (unless we would want to change the internal representation on read).
When `Frequencies` is accessed we could either return a reference (if we fix the representation, but the interface would be non-uniform), or also return by value. In that case there is no actual copy of the data, since `Frequencies` is a wrapper around `cow_ptr<std::vector<double>>`.